compiler/rustc_parse/src/validate_attr.rs

```diff
@@ -167,7 +167,20 @@ pub fn check_attribute_safety(psess: &ParseSess, safety: AttributeSafety, attr:
             // square bracket respectively.
             let diag_span = attr_item.span();
 
-            if attr.span.at_least_rust_2024() {
+            // Attributes that were safe before edition 2024, but are unsafe afterwards.
+            const SAFE_BEFORE_RUST_2024: &[Symbol] =
+                &[sym::export_name, sym::link_section, sym::no_mangle];
```

Comment:
Instead of hard-coding this, you could modify `AttributeSafety`. For example, make its `Unsafe` variant hold `{ since: Option<Edition> }`. However, I don't know how large the ripple effect would be in terms of LOCs.

Reply:
On a second look, it should be quite easy to do. You just need to modify the "`unsafe`" macro matcher of macro `gated` in mod `builtin_attrs` to optionally take an edition, maybe via `unsafe $(($edition:ident..))?`, so we can mark these attributes as `unsafe(Rust2024..)` (or whatever other syntax seems more legible to you).

Reply:
allright, gave that a go. I don't think using `$(....)?` is really doable here, because you can't handle the case where the value is unspecified well (there are workarounds, but I think here just writing out an extra case is simpler).

Reply:
Fair. I mean you'd just need to create two helper macro *rules* but it's okay to delay that until we can no longer deal with the combinatorial explosion.